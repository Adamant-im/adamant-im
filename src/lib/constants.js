export const EPOCH = Date.UTC(2017, 8, 2, 17, 0, 0, 0)

export const Transactions = {
  SEND: 0,
  SIGNATURE: 1,
  DELEGATE: 2,
  VOTE: 3,
  MULTI: 4,
  DAPP: 5,
  IN_TRANSFER: 6,
  OUT_TRANSFER: 7,
  CHAT_MESSAGE: 8,
  STATE: 9
}

export const Cryptos = {
  ADM: 'ADM',
  ETH: 'ETH',
  BNB: 'BNB',
  DOGE: 'DOGE',
  BZ: 'BZ'
}

export const ERC20 = Object.freeze([
  Cryptos.BNB,
  Cryptos.BZ
])

export const isErc20 = crypto => ERC20.includes(crypto)

/** Number of decimal places for the different crypto amounts */
export const CryptoAmountPrecision = {
  ADM: 2,
  ETH: 6,
  BNB: 6,
  DOGE: 8,
  BZ: 6
}

/** Fees for the misc ADM operations */
export const Fees = {
  /** Storing a value into the KVS */
  KVS: 0.001,
  /** Transfering tokens */
  ADM_TRANSFER: 0.5,
  NOT_ADM_TRANSFER: 0.001
}

/** Regex for detecting of base64 encoded string */
export const base64regex = /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/

export const Symbols = {
  CLOCK: String.fromCharCode(0x23f0), // ⏰
  HOURGLASS: String.fromCharCode(0x23f3), // ⏳
  CROSS: String.fromCharCode(0x274c) // ❌
}

export const Delegates = {
  ACTIVE_DELEGATES: 101
}

/** Gas value for the ETH transfers */
export const ETH_TRANSFER_GAS = 21000

/** Gas value for the ERC-20 transfers */
export const ERC20_TRANSFER_GAS = ETH_TRANSFER_GAS * 10

export default {
  EPOCH,
  Transactions
}

export const UserPasswordArticleLink = 'https://medium.com/adamant-im/more-convenience-login-to-the-web-messenger-with-user-password-9d48a736dfd8'

export const UserPasswordHashSettings = {
  SALT: 'salt',
  ITERATIONS: 100000,
  KEYLEN: 64,
  DIGEST: 'sha512'
}
