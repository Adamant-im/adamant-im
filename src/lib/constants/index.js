import { Cryptos, CryptosInfo, CryptosOrder } from './cryptos'

export const EPOCH = Date.UTC(2017, 8, 2, 17, 0, 0, 0)

export const Transactions = {
  SEND: 0,
  SIGNATURE: 1,
  DELEGATE: 2,
  VOTE: 3,
  MULTI: 4,
  DAPP: 5,
  IN_TRANSFER: 6,
  OUT_TRANSFER: 7,
  CHAT_MESSAGE: 8,
  STATE: 9
}

/**
 * @see https://github.com/Adamant-im/adamant/wiki/Message-Types
 */
export const MessageType = {
  BASIC_ENCRYPTED_MESSAGE: 1,
  RICH_CONTENT_MESSAGE: 2,
  cryptoTransferMessage: (cryptoSymbol) => `${cryptoSymbol.toLowerCase()}_transaction`
}

export const Rates = {
  USD: 'USD',
  EUR: 'EUR',
  RUB: 'RUB',
  CNY: 'CNY',
  JPY: 'JPY'
}

export const RatesNames = {
  [Rates.USD]: 'USD ($)', // United States Dollar
  [Rates.EUR]: 'EUR (‚Ç¨)', // Euro
  [Rates.RUB]: 'RUB (‚ÇΩ)', // Russian Ruble
  [Rates.CNY]: 'CNY (¬•)', // Chinese Yuan
  [Rates.JPY]: 'JPY (¬•)' // Japanese Yen
}

/** Fees for the misc ADM operations */
export const Fees = {
  /** Storing a value into the KVS */
  KVS: 0.001,
  /** Transferring tokens */
  ADM_TRANSFER: 0.5,
  NOT_ADM_TRANSFER: 0.001
}

/** Regex for detecting of base64 encoded string */
export const base64regex =
  /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/

export const Symbols = {
  CLOCK: String.fromCharCode(0x23f0), // ‚è∞
  HOURGLASS: String.fromCharCode(0x23f3), // ‚è≥
  CROSS: String.fromCharCode(0x274c) // ‚ùå
}

export const Delegates = {
  ACTIVE_DELEGATES: 101
}

export const WelcomeMessage = {
  ADAMANT_BOUNTY: 'ADAMANT Bounty',
  ADAMANT_ICO: 'ADAMANT Tokens'
}

export const BTC_BASED = Object.freeze([Cryptos.DOGE, Cryptos.DASH, Cryptos.BTC])

export const LSK_BASED = Object.freeze([Cryptos.LSK])

export const INSTANT_SEND = Object.freeze([Cryptos.DASH])

// Some cryptos allows to save public data with a Tx
export const ALLOW_TEXT_DATA = Object.freeze([Cryptos.LSK])

export const isErc20 = (crypto) => CryptosInfo[crypto].type === 'ERC20'

export const isEthBased = (crypto) => isErc20(crypto) || crypto === Cryptos.ETH

export const isFeeEstimate = (crypto) => isEthBased(crypto)

export const isBtcBased = (crypto) => BTC_BASED.includes(crypto)

export const isLskBased = (crypto) => LSK_BASED.includes(crypto)

export const isSelfTxAllowed = (crypto) => LSK_BASED.includes(crypto) || crypto === Cryptos.ADM

export const isInstantSendPossible = (crypto) => INSTANT_SEND.includes(crypto)

export const isTextDataAllowed = (crypto) => ALLOW_TEXT_DATA.includes(crypto)

export const RE_LSK_ADDRESS_LEGACY = /^[0-9]{2,21}L$/

/**
 * These gas limit values are used only for estimate fees for ETH & ERC-20 transfers in the Send tokens form,
 * Actual gas limit values are calculated with estimateGas(transactionObject)
 * when each specific transaction is created
 */

/** Gas limit value for the ETH transfers */
export const DEFAULT_ETH_TRANSFER_GAS = CryptosInfo['ETH'].defaultGasLimit
/** Gas limit value for the ERC-20 transfers */
export const DEFAULT_ERC20_TRANSFER_GAS = DEFAULT_ETH_TRANSFER_GAS * 2.4

/** Increase fee multiplier. Used as a checkbox on SendFundsForm */
export const INCREASE_FEE_MULTIPLIER = 2

export { Cryptos, CryptosInfo, CryptosOrder }

export default {
  EPOCH,
  Transactions
}

export const UserPasswordArticleLink =
  'https://medium.com/adamant-im/more-convenience-login-to-the-web-messenger-with-user-password-9d48a736dfd8'

export const UserPasswordHashSettings = {
  SALT: 'salt',
  ITERATIONS: 100000,
  KEYLEN: 64,
  DIGEST: 'sha512'
}

/** Status of ADM or coin transaction */
export const TransactionStatus = {
  CONFIRMED: 'CONFIRMED', // Tx has at least 1 network confirmation
  PENDING: 'PENDING', // We don't know about this Tx anything yet, it may not exist in a blockchain
  REGISTERED: 'REGISTERED', // This Ts is seen on a blockchain, but has 0 confirmations yet
  REJECTED: 'REJECTED', // We unable to find this Tx in a blockchain, or it says it was rejected
  INVALID: 'INVALID', // Wrong sender, recipient, time or amount
  UNKNOWN: 'UNKNOWN' // We don't recognize a cryptocurrency
}

/** Additional status of ADM or coin transaction, which leads to new virtualStatus */
export const TransactionAdditionalStatus = {
  NONE: false,
  INSTANT_SEND: 'instant_send', // Dash InstantSend enabled transaction
  ADM_REGISTERED: 'adm_registered' // ADM tx, registered in a blockchain, but has 0 confirmations yet
}

export const tsIcon = function (status) {
  if (status === TransactionStatus.CONFIRMED) {
    return 'mdi-check'
  } else if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return 'mdi-clock-outline'
  } else if (status === TransactionStatus.REJECTED) {
    return 'mdi-close-circle-outline'
  } else if (status === TransactionStatus.INVALID) {
    return 'mdi-alert-outline'
  } else if (status === TransactionStatus.UNKNOWN) {
    return 'mdi-help-circle-outline'
  }
}

export const tsColor = function (status) {
  if (status === TransactionStatus.REJECTED) {
    return 'red'
  } else if (status === TransactionStatus.INVALID || status === TransactionStatus.UNKNOWN) {
    return 'yellow'
  }
  return ''
}

export const tsUpdatable = function (status, currency) {
  currency = currency.toUpperCase()
  if (currency === Cryptos.ADM) {
    return false
  } else if (status === TransactionStatus.CONFIRMED) {
    return true
  } else if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return false
  } else if (status === TransactionStatus.REJECTED) {
    return true
  } else if (status === TransactionStatus.INVALID) {
    return true
  } else if (status === TransactionStatus.UNKNOWN) {
    return false
  }
}

/**
 * Returns minimal amount that can be transferred for the specified crypto
 * @param {string} crypto crypto
 * @returns {number}
 */
export function getMinAmount(crypto) {
  let amount = CryptosInfo[crypto].minTransferAmount

  if (!amount) {
    const precision = CryptosInfo[crypto].cryptoTransferDecimals
    amount = precision ? Math.pow(10, -precision) : 0
  }

  return amount
}

/**
 * When clicking on replied message inside the chat that animation duration will
 * be applied while scrolling.
 **/
export const SCROLL_TO_REPLIED_MESSAGE_ANIMATION_DURATION = 300 // ms

export const FetchStatus = {
  Idle: 'idle',
  Loading: 'loading',
  Success: 'success',
  Error: 'error'
}

export const REACT_EMOJIS = {
  FACE_WITH_TEARS_OF_JOY: 'üòÇ',
  FIRE: 'üî•',
  GRINNING_FACE_WITH_SMILING_EYES: 'üòÑ',
  THUMB_UP: 'üëç',
  OK_HAND: 'üëå',
  RED_HEART: '‚ù§Ô∏è',
  SLIGHTLY_SMILING_FACE: 'üôÇ',
  THINKING_FACE: 'ü§î',
  WAVING_HAND: 'üëã',
  FOLDED_HANDS: 'üôè',
  FLUSHED_FACE: 'üò≥',
  PARTY_POPPER: 'üéâ'
}
