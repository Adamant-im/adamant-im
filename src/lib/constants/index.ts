import {
  AllCryptos,
  AllCryptosOrder,
  Cryptos,
  CryptosInfo,
  CryptosOrder,
  CryptoSymbol
} from './cryptos'

import { mdiCheck, mdiClockOutline, mdiCloseCircleOutline, mdiAlertOutline, mdiHelpCircleOutline } from '@mdi/js'


export const EPOCH = Date.UTC(2017, 8, 2, 17, 0, 0, 0)

export const Transactions = {
  SEND: 0,
  SIGNATURE: 1,
  DELEGATE: 2,
  VOTE: 3,
  MULTI: 4,
  DAPP: 5,
  IN_TRANSFER: 6,
  OUT_TRANSFER: 7,
  CHAT_MESSAGE: 8,
  STATE: 9
} as const

/**
 * @see https://github.com/Adamant-im/adamant/wiki/Message-Types
 */
export const MessageType = {
  BASIC_ENCRYPTED_MESSAGE: 1,
  RICH_CONTENT_MESSAGE: 2,
  SIGNAL_MESSAGE: 3,
  cryptoTransferMessage: (cryptoSymbol: string) => `${cryptoSymbol.toLowerCase()}_transaction`
} as const

export type FiatCurrencySymbol = 'USD' | 'EUR' | 'RUB' | 'CNY' | 'JPY'

export const Rates: Record<FiatCurrencySymbol, FiatCurrencySymbol> = {
  USD: 'USD',
  EUR: 'EUR',
  RUB: 'RUB',
  CNY: 'CNY',
  JPY: 'JPY'
}

export const RatesNames: Record<FiatCurrencySymbol, string> = {
  USD: 'USD ($)', // United States Dollar
  EUR: 'EUR (‚Ç¨)', // Euro
  RUB: 'RUB (‚ÇΩ)', // Russian Ruble
  CNY: 'CNY (¬•)', // Chinese Yuan
  JPY: 'JPY (¬•)' // Japanese Yen
}

export const UPLOAD_MAX_FILE_COUNT = 5
export const UPLOAD_MAX_FILE_SIZE = 10 * 1024 * 1024 // 10MB

/** Fees for the misc ADM operations */
export const Fees = {
  /** Storing a value into the KVS */
  KVS: 0.001,
  /** Transferring tokens */
  ADM_TRANSFER: 0.5,
  NOT_ADM_TRANSFER: 0.001
}

/** Regex for detecting of base64 encoded string */
export const base64regex =
  /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/

export const Symbols = {
  CLOCK: String.fromCharCode(0x23f0), // ‚è∞
  HOURGLASS: String.fromCharCode(0x23f3), // ‚è≥
  CROSS: String.fromCharCode(0x274c) // ‚ùå
}

export const Delegates = {
  ACTIVE_DELEGATES: 101
}

export const WelcomeMessage = {
  ADAMANT_BOUNTY: 'ADAMANT Bounty',
  ADAMANT_ICO: 'ADAMANT Tokens'
}

export const BTC_BASED = Object.freeze([Cryptos.DOGE, Cryptos.DASH, Cryptos.BTC])
export const KLY_BASED = Object.freeze([Cryptos.KLY])
export const INSTANT_SEND = Object.freeze([Cryptos.DASH])
export const ALLOW_TEXT_DATA = Object.freeze([Cryptos.KLY]) // Some blockchains allow storing a text message within a Tx

export const isErc20 = (crypto: CryptoSymbol) => CryptosInfo[crypto]?.type === 'ERC20'
export const isEthBased = (crypto: CryptoSymbol) => isErc20(crypto) || crypto === Cryptos.ETH
export const isFeeEstimate = (crypto: CryptoSymbol) => isEthBased(crypto)
export const isBtcBased = (crypto: CryptoSymbol) => BTC_BASED.includes(crypto)
export const isKlyBased = (crypto: CryptoSymbol) => KLY_BASED.includes(crypto)
export const isSelfTxAllowed = (crypto: CryptoSymbol) =>
  KLY_BASED.includes(crypto) || crypto === Cryptos.ADM
export const isInstantSendPossible = (crypto: CryptoSymbol) => INSTANT_SEND.includes(crypto)
export const isTextDataAllowed = (crypto: CryptoSymbol) => ALLOW_TEXT_DATA.includes(crypto)

export const RE_KLY_ADDRESS_LEGACY = /^[0-9]{2,21}L$/

/**
 * These gas limit values are used only for estimate fees for ETH & ERC-20 transfers in the Send tokens form,
 * Actual gas limit values are calculated with estimateGas(transactionObject)
 * when each specific transaction is created
 */

/** Gas limit value for the ETH transfers */
export const DEFAULT_ETH_TRANSFER_GAS_LIMIT = (CryptosInfo['ETH'] as any).defaultGasLimit // @todo fix type in adamant-wallets
/** Gas limit value for the ERC-20 transfers */
export const DEFAULT_ERC20_TRANSFER_GAS_LIMIT = DEFAULT_ETH_TRANSFER_GAS_LIMIT * 2.4

/** Increase fee multiplier. Used as a checkbox on SendFundsForm */
export const INCREASE_FEE_MULTIPLIER = 1.5

export { AllCryptos, AllCryptosOrder, Cryptos, CryptosInfo, CryptosOrder, type CryptoSymbol }

export default {
  EPOCH,
  Transactions
}

export const UserPasswordArticleLink =
  'https://medium.com/adamant-im/more-convenience-login-to-the-web-messenger-with-user-password-9d48a736dfd8'

export const UserPasswordHashSettings = {
  SALT: 'salt',
  ITERATIONS: 100000,
  KEYLEN: 64,
  DIGEST: 'sha512'
}

export type TransactionStatusType =
  | 'CONFIRMED'
  | 'PENDING'
  | 'REGISTERED'
  | 'REJECTED'
  | 'INVALID'
  | 'UNKNOWN'

/** Status of ADM or coin transaction */
export const TransactionStatus: Record<TransactionStatusType, TransactionStatusType> = {
  CONFIRMED: 'CONFIRMED', // Tx has at least 1 network confirmation
  PENDING: 'PENDING', // We don't know about this Tx anything yet, it may not exist in a blockchain
  REGISTERED: 'REGISTERED', // This Ts is seen on a blockchain, but has 0 confirmations yet
  REJECTED: 'REJECTED', // We unable to find this Tx in a blockchain, or it says it was rejected
  INVALID: 'INVALID', // Wrong sender, recipient, time or amount
  UNKNOWN: 'UNKNOWN' // We don't recognize a cryptocurrency
}

/** Additional status of ADM or coin transaction, which leads to new virtualStatus */
export const TransactionAdditionalStatus = {
  NONE: false,
  INSTANT_SEND: 'instant_send', // Dash InstantSend enabled transaction
  ADM_REGISTERED: 'adm_registered' // ADM tx, registered in a blockchain, but has 0 confirmations yet
}

export const tsIcon = function (status: TransactionStatusType) {
  if (status === TransactionStatus.CONFIRMED) {
    return mdiCheck
  }

  if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return mdiClockOutline
  }

  if (status === TransactionStatus.REJECTED) {
    return mdiCloseCircleOutline
  }

  if (status === TransactionStatus.INVALID) {
    return mdiAlertOutline
  }

  if (status === TransactionStatus.UNKNOWN) {
    return mdiHelpCircleOutline
  }
}

export const tsColor = function (status: TransactionStatusType) {
  if (status === TransactionStatus.REJECTED) {
    return 'red'
  } else if (status === TransactionStatus.INVALID || status === TransactionStatus.UNKNOWN) {
    return 'yellow'
  }
  return ''
}

export const tsUpdatable = function (status: TransactionStatusType, currency: CryptoSymbol) {
  currency = currency.toUpperCase() as CryptoSymbol

  if (currency === Cryptos.ADM) {
    return false
  } else if (status === TransactionStatus.CONFIRMED) {
    return true
  } else if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return false
  } else if (status === TransactionStatus.REJECTED) {
    return true
  } else if (status === TransactionStatus.INVALID) {
    return true
  } else if (status === TransactionStatus.UNKNOWN) {
    return false
  }
}

/**
 * Returns minimal amount that can be transferred for the specified crypto
 * @param {string} crypto crypto
 * @returns {number}
 */
export function getMinAmount(crypto: CryptoSymbol) {
  let amount = CryptosInfo[crypto].minTransferAmount

  if (!amount) {
    const precision = CryptosInfo[crypto].cryptoTransferDecimals
    amount = precision ? Math.pow(10, -precision) : 0
  }

  return amount
}

/**
 * When clicking on replied message inside the chat that animation duration will
 * be applied while scrolling.
 **/
export const SCROLL_TO_REPLIED_MESSAGE_ANIMATION_DURATION = 300 // ms

export const FetchStatus = {
  Idle: 'idle',
  Loading: 'loading',
  Success: 'success',
  Error: 'error'
} as const

export const REACT_EMOJIS = {
  FACE_WITH_TEARS_OF_JOY: 'üòÇ',
  FIRE: 'üî•',
  GRINNING_FACE_WITH_SMILING_EYES: 'üòÑ',
  THUMB_UP: 'üëç',
  OK_HAND: 'üëå',
  RED_HEART: '‚ù§Ô∏è',
  SLIGHTLY_SMILING_FACE: 'üôÇ',
  THINKING_FACE: 'ü§î',
  WAVING_HAND: 'üëã',
  FOLDED_HANDS: 'üôè',
  FLUSHED_FACE: 'üò≥',
  PARTY_POPPER: 'üéâ'
} as const
