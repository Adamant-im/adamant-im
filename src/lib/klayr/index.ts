export * from './types/klayr'
export * from './klayr-constants'
export * from './klayr-account'
