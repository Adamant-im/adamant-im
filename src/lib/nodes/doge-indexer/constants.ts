export const NB_BLOCKS = 5 // Number of last blocks for fee estimation
