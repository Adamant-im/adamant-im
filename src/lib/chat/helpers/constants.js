import { Cryptos } from '@/lib/constants'

/**
 * Maps crypto transfer transaction types to the respective cryptos.
 * E.g.:
 *
 * ```
 * {
 *   'eth_transaction': 'ETH',
 *   'bnb_transaction': 'BNB'
 * }
 * ```
 */
export const KnownCryptos = Object.keys(Cryptos).reduce((map, crypto) => {
  if (crypto !== Cryptos.ADM) {
    const key = `${crypto.toLowerCase()}_transaction`
    map[key] = crypto
  }
  return map
}, {})

/** Cryptos, supported by other clients, but not PWA */
export const UnsupportedCryptos = {
  lsk_transaction: 'LSK'
}
