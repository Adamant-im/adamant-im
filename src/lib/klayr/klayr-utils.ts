/**
 * Additional coin-based functions here
 */

import { CryptoSymbol } from '@/lib/constants/cryptos'
import { TRANSACTION_PARAMS_SCHEMA, TRANSACTION_SCHEMA } from '@/lib/klayr/klayr-schemas'
import { DecodedTransaction, Transaction } from './types/klayr'
import { codec } from '@klayr/codec'
import * as cryptography from '@klayr/cryptography'
import networks from '@/lib/klayr/networks'
import { convertBeddowsTokly } from '@klayr/transactions'
import * as transactions from '@klayr/transactions'
import { Buffer } from 'buffer'
import pbkdf2 from 'pbkdf2'
import sodium from 'sodium-browserify-tweetnacl'
import {
  KlayrHashSettings,
  KLY_CHAIN_ID,
  KLY_TRANSFER_TO_NEW_ACCOUNT_FEE,
  KLY_DECIMALS,
  KLY_DEMO_ACCOUNT,
  KLY_TOKEN_ID,
  KLY_MIN_FEE_PER_BYTE
} from './klayr-constants'
import { bytesToHex, hexToBytes } from '@/lib/hex'
import * as validator from '@klayr/validator'

/**
 * Returns Millis timestamp by KLY UNIX timestamp (sec)
 * @param {number} timestamp
 */
export function getMillisTimestamp(timestamp: number | string) {
  return parseInt(timestamp.toString()) * 1000
}

/**
 * Returns KLY timestamp (UNIX in sec) by Millis timestamp
 * @param {number} millisTimestamp
 */
export function getKlayrTimestamp(millisTimestamp: number | string) {
  return Math.round(parseInt(millisTimestamp.toString()) / 1000) // may be a mistake (use Math.floor instead)
}

export function getAccount(crypto: CryptoSymbol, passphrase: string) {
  const network = networks[crypto]
  const klayrSeed = pbkdf2.pbkdf2Sync(
    passphrase,
    KlayrHashSettings.SALT,
    KlayrHashSettings.ITERATIONS,
    KlayrHashSettings.KEYLEN,
    KlayrHashSettings.DIGEST
  )
  const keyPair = sodium.crypto_sign_seed_keypair(klayrSeed)
  const addressHexBinary = cryptography.address.getAddressFromPublicKey(keyPair.publicKey)
  const addressHex = bytesToHex(addressHexBinary)
  const address = cryptography.address.getKlayr32AddressFromPublicKey(keyPair.publicKey)
  // Don't work currently https://github.com/LiskHQ/lisk-sdk/issues/6651
  // const addressLegacy = cryptography.getLegacyAddressFromPublicKey(keyPair.publicKey)
  // const addressLegacy = cryptography.getLegacyAddressFromPrivateKey(keyPair.secretKey)
  const addressLegacy = 'cryptography.getLegacyAddressFromPublicKey(bytesToHex(keyPair.publicKey))'
  return {
    network,
    keyPair,
    address,
    addressHexBinary,
    addressHex,
    addressLegacy
  }
}

export function seedKlyAccount(passphrase: string) {
  const klayrSeed = pbkdf2.pbkdf2Sync(
    passphrase,
    KlayrHashSettings.SALT,
    KlayrHashSettings.ITERATIONS,
    KlayrHashSettings.KEYLEN,
    KlayrHashSettings.DIGEST
  )
  const keyPair = sodium.crypto_sign_seed_keypair(klayrSeed)

  const address = cryptography.address.getKlayr32AddressFromPublicKey(keyPair.publicKey)
  const addressHexBinary = cryptography.address.getAddressFromPublicKey(keyPair.publicKey)
  const addressHex = bytesToHex(addressHexBinary)

  return {
    keyPair,
    address,
    addressHexBinary,
    addressHex
  }
}

/**
 * Create and sign a transfer transaction.
 * Transaction `hex` result is ready for broadcasting to blockchain network.
 *
 * @param keyPair Sender's `publicKey` and `privateKey`
 * @param address Receiver address in Klayr32 format
 * @param amount Amount to transfer (KLY, not Beddows)
 * @param fee Transaction fee (KLY, not Beddows)
 * @param nonce Transaction nonce
 * @param data Transaction data field
 */
export function createTransaction(
  keyPair: { publicKey: Buffer; secretKey: Buffer },
  address: string,
  amount: number | string,
  fee: number | string,
  nonce: number | string,
  data = ''
) {
  const unsignedTransaction = createUnsignedTransaction(
    address,
    Buffer.from(keyPair.publicKey.buffer).toString('hex'),
    amount,
    fee,
    nonce,
    data
  )

  const signedTransaction = transactions.signTransaction(
    unsignedTransaction,
    Buffer.from(KLY_CHAIN_ID, 'hex'),
    keyPair.secretKey,
    TRANSACTION_PARAMS_SCHEMA
  ) as unknown as Transaction

  const id = bytesToHex(signedTransaction.id)

  return { hex: encodeTransaction(signedTransaction), id }
}

/**
 * Creates unsigned KLY transaction.
 *
 * @param recipientAddress Recipient address (Klayr32 string hex format)
 * @param senderPublicKey Sender public key (string hex)
 * @param amount Amount of KLY to send
 * @param fee Transaction fee
 * @param nonce Last nonce
 * @param data A.k.a note
 */
export function createUnsignedTransaction(
  recipientAddress: string,
  senderPublicKey: string,
  amount: number | string,
  fee: number | string,
  nonce: number | string,
  data = ''
) {
  const amountString = transactions.convertklyToBeddows((+amount).toFixed(KLY_DECIMALS))
  const feeString = transactions.convertklyToBeddows((+fee).toFixed(KLY_DECIMALS))
  const nonceString = nonce.toString()

  // Adjust the values of the unsigned transaction manually
  const unsignedTransaction = {
    module: 'token',
    command: 'transfer',
    fee: BigInt(feeString),
    nonce: BigInt(nonceString),
    senderPublicKey: Buffer.from(senderPublicKey, 'hex'),
    params: Buffer.alloc(0) as unknown, // must be Buffer when passing to `validator.validate()` and then overridden with an object
    signatures: []
  }

  validator.validator.validate(TRANSACTION_SCHEMA, unsignedTransaction)

  // Create the asset for the Token Transfer transaction
  const transferParams = {
    tokenID: Buffer.from(KLY_TOKEN_ID, 'hex'),
    amount: BigInt(amountString),
    recipientAddress: cryptography.address.getAddressFromKlayr32Address(recipientAddress),
    data
  }

  // Add the transaction params to the transaction object
  unsignedTransaction.params = transferParams

  // Only `transaction.params` will be validated
  transactions.validateTransaction(unsignedTransaction, TRANSACTION_PARAMS_SCHEMA)

  return unsignedTransaction
}

/**
 * Encode transaction for later broadcasting to a node.
 */
export function encodeTransaction(transaction: DecodedTransaction | Transaction) {
  let encodedParams
  if (!Buffer.isBuffer(transaction.params)) {
    encodedParams = codec.encode(TRANSACTION_PARAMS_SCHEMA, transaction.params)
  } else {
    encodedParams = transaction.params
  }

  const encodedTransaction = codec.encode(TRANSACTION_SCHEMA, {
    ...transaction,
    params: encodedParams
  })

  return encodedTransaction.toString('hex')
}

type EstimateFeeParams = {
  /**
   * KLY amount
   */
  amount?: string | number
  /**
   * Sender's `publicKey` and `privateKey`
   */
  keyPair?: { publicKey: string; secretKey: string }
  /**
   * Recipient address in Klayr32 format
   */
  recipientAddress?: string
  /**
   * Transaction data field
   */
  data?: string
  isNewAccount?: boolean
  /**
   * Current nonce
   */
  nonce?: number | string
}

/**
 * Estimate transaction fee by KLY amount.
 * https://klayr.xyz/documentation/understand-blockchain/blocks-txs.html#transaction-fees
 *
 * @param params Transaction params
 */
export function estimateFee(params?: EstimateFeeParams) {
  const {
    amount = '1',
    keyPair = KLY_DEMO_ACCOUNT.keyPair,
    recipientAddress = KLY_DEMO_ACCOUNT.address,
    data = '',
    nonce = 0,
    isNewAccount
  } = params || {}

  const transaction = createTransaction(
    {
      publicKey: Buffer.from(keyPair.publicKey, 'hex'),
      secretKey: Buffer.from(keyPair.secretKey, 'hex')
    },
    recipientAddress,
    amount,
    1,
    nonce,
    data
  )
  const transactionBytes = hexToBytes(transaction.hex)

  const fee = BigInt(transactionBytes.length) * KLY_MIN_FEE_PER_BYTE
  const transferToNewAccountFee = isNewAccount ? KLY_TRANSFER_TO_NEW_ACCOUNT_FEE : BigInt(0)

  const totalFee = fee + transferToNewAccountFee

  return convertBeddowsTokly(totalFee.toString())
}
