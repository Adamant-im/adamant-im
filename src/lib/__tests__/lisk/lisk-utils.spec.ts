// @vitest-environment node
// Some crypto libs throw errors when using `jsdom` environment

import { KLY_MIN_FEE_PER_BYTE } from '@/lib/klayr'
import { convertBeddowsTokly } from '@klayr/transactions'
import { describe, it, expect } from 'vitest'
import { Cryptos } from '@/lib/constants'
import { estimateFee, getAccount } from '@/lib/klayr/klayr-utils'

const passphrase = 'joy mouse injury soft decade bid rough about alarm wreck season sting'

describe('lisk-utils', () => {
  describe('getAccount', () => {
    it('should generate account from passphrase', () => {
      expect(getAccount(Cryptos.KLY, passphrase)).toMatchSnapshot()
    })
  })

  describe('estimateFee', () => {
    it('should calculate fee including `data`', () => {
      const data = 'hello'
      const minimalFee = BigInt(165000)
      const messageFee = BigInt(data.length) * BigInt(KLY_MIN_FEE_PER_BYTE)
      const expectedFee = minimalFee + messageFee

      expect(estimateFee({ data })).toBe(convertBeddowsTokly(expectedFee.toString()))
    })
  })
})
