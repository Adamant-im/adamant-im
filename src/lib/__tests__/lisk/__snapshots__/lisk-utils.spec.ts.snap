// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`klayr-utils > getAccount > should generate account from passphrase 1`] = `
{
  "address": "klykjurzk3xb47scma49ukyqupn8vrg2ggyuehk5j",
  "addressHex": "a721205101325db197296c291ac08778a5fc7ff1",
  "addressHexBinary": {
    "data": [
      167,
      33,
      32,
      81,
      1,
      50,
      93,
      177,
      151,
      41,
      108,
      41,
      26,
      192,
      135,
      120,
      165,
      252,
      127,
      241,
    ],
    "type": "Buffer",
  },
  "addressLegacy": "cryptography.getLegacyAddressFromPublicKey(bytesToHex(keyPair.publicKey))",
  "keyPair": {
    "publicKey": {
      "data": [
        50,
        235,
        138,
        255,
        40,
        214,
        53,
        186,
        3,
        178,
        187,
        108,
        199,
        133,
        7,
        44,
        12,
        1,
        18,
        58,
        36,
        245,
        110,
        88,
        15,
        249,
        199,
        50,
        6,
        17,
        241,
        29,
      ],
      "type": "Buffer",
    },
    "secretKey": {
      "data": [
        252,
        163,
        131,
        181,
        56,
        88,
        159,
        151,
        39,
        66,
        203,
        135,
        3,
        62,
        247,
        137,
        79,
        255,
        122,
        123,
        109,
        82,
        199,
        75,
        221,
        247,
        228,
        247,
        248,
        102,
        10,
        13,
        50,
        235,
        138,
        255,
        40,
        214,
        53,
        186,
        3,
        178,
        187,
        108,
        199,
        133,
        7,
        44,
        12,
        1,
        18,
        58,
        36,
        245,
        110,
        88,
        15,
        249,
        199,
        50,
        6,
        17,
        241,
        29,
      ],
      "type": "Buffer",
    },
  },
  "network": {
    "name": "Klayr",
    "port": 8000,
    "unit": "KLY",
    "wsPort": 8001,
  },
}
`;
